### PR TITLE
WIP: Made PR changes, should I rely on infinite board? 

### PR DESCRIPTION
**Pull request by ester-agishtein**

### CODE REVIEW

File: src/main/java/touro/snake/AIKeyListener.java [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+
+import java.awt.event.KeyAdapter;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.Random;
+import java.awt.event.KeyAdapter;
+import java.awt.event.KeyEvent;
+
+public class AIKeyListener extends KeyAdapter {
```
Comment (andrewoid) [inline src/main/java/touro/snake/AIKeyListener.java:11]: Copy and pasting code without understanding the code you are copying is highly dangerous.

What does a KeyAdapter do? Why is this a KeyAdapter?

**Comment by andrewoid** [comment]
Perhaps you should use the debugger and try to figure out why the program doesn't load...

File: src/main/java/touro/snake/AIKeyListener.java (hunk 1 above)
Comment (ester-agishtein) [reply src/main/java/touro/snake/AIKeyListener.java:11]: I originally wasn't doing this and just trying to make some "change directions" methods but was having trouble getting the snake to show up. I was trying to keep as close to the original snake's work flow as possible but now realized my mistake and changing tactic.

File: src/main/java/touro/snake/AIKeyListener.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
+package touro.snake;
+
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.Random;
+
+public class AIKeyListener{
```
Comment (andrewoid) [inline src/main/java/touro/snake/AIKeyListener.java:8]: Making sure code makes sense is important. This isn't a KeyListener so don't name it KeyListener

File: src/test/java/touro/snake/SnakeTest.java [hunk 3]
```diff
@@ -131,7 +131,7 @@ public void inBounds_true() {
     public void inBounds_false() {
         //given
         SnakeHeadStateMachine snakeHeadStateMachine = mock(SnakeHeadStateMachine.class);
-        Snake snake = new Snake(snakeHeadStateMachine);
+        Snake snake = mock(Snake.class);
```
Comment (andrewoid) [inline src/test/java/touro/snake/SnakeTest.java:134]: You cannot test a mock.
If you are testing a Snake then it needs to be an actual Snake class

File: src/main/java/touro/snake/AISnakeCreator.java [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    public void moveSelf(Snake aiSnake) throws InterruptedException {
+        List<String> directions = getDirections();
+        int len = directions.size();
+
+        if(aiSnake.inBounds() && !aiSnake.eatsSelf()){
+            int randIndex = rand.nextInt(len);
+            String dirString = directions.get(randIndex);
+            Direction direction = Direction.valueOf(dirString);
+            aiSnake.turnTo(direction);
+            Thread.sleep(DELAY_MS);
```
Comment (andrewoid) [inline src/main/java/touro/snake/AISnakeCreator.java:27]: Why are you calling Thread.sleep?

Comment (ester-agishtein) [reply src/main/java/touro/snake/AISnakeCreator.java:27]: I wanted to create a delay so the moving of the snake feels more natural

File: src/main/java/touro/snake/GardenThread.java [hunk 5]
```diff
@@ -18,9 +18,12 @@ public GardenThread(Garden garden, GardenView gardenView) {
      * Every second, advance the Garden (moving the Snake, eating, death) and repaint the Garden
      */
     public void run() {
-        while (garden.advance()) {
-            gardenView.repaint();
+        while (garden.advance(garden.getSnake()) && garden.aiAdvance(garden.getAiSnake())) {
             try {
+                gardenView.repaint();
+                garden.advance(garden.getAiSnake());
+                AISnakeCreator aiSnakeCreator = new AISnakeCreator(garden.getAiSnake());
```
Comment (andrewoid) [inline src/main/java/touro/snake/GardenThread.java:25]: Why are you recreating `AISnakeCreator` each time?

File: src/main/java/touro/snake/Direction.java [hunk 6]
```diff
@@ -4,5 +4,11 @@
  * The possible directions that the Snake's head can be in.
  */
 public enum Direction {
-    North, East, South, West
+    North("North"), East("East"), South("South"), West("West");
```
Comment (andrewoid) [inline src/main/java/touro/snake/Direction.java:7]: The "value" is not necessary.
`North.toString()` will give you the same String you're returning without it.

File: src/main/java/touro/snake/AISnakeCreator.java (hunk 4 above)
Comment (andrewoid) [reply src/main/java/touro/snake/AISnakeCreator.java:27]: The game already updates every DELAY_MS. You are adding to the delay.

File: src/main/java/touro/snake/AISnakeCreator.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        moveSelf(aiSnake);
+    }
+
+    public void moveSelf(Snake aiSnake) throws InterruptedException {
+        List<String> directions = getDirections();
+        int len = directions.size();
+
+        if(aiSnake.inBounds() && !aiSnake.eatsSelf()){
+            int randIndex = rand.nextInt(len);
+            String dirString = directions.get(randIndex);
```
Comment (andrewoid) [inline src/main/java/touro/snake/AISnakeCreator.java:24]: You are turning `Direction.values()` into a `List<String>`, to get random Direction (as a String), then turning that String back into a Direction.

You should be able to do this without turning them into Strings.

File: src/main/java/touro/snake/Garden.java [hunk 8]
```diff
@@ -31,37 +38,49 @@ public Food getFood() {
      *
      * @return true if the snake is still alive, otherwise false.
      */
-    public boolean advance() {
-        if (moveSnake()) {
+    public boolean advance(Snake snake) {
```
Comment (andrewoid) [inline src/main/java/touro/snake/Garden.java:41]: I'd rather you move both Snakes in the one `advance()` method.

**Review by andrewoid — CHANGES_REQUESTED** [review]
Lot of changes necessary.

File: src/main/java/touro/snake/GardenThread.java (hunk 5 above)
Comment (ester-agishtein) [reply src/main/java/touro/snake/GardenThread.java:25]: My bad, will move.

File: src/main/java/touro/snake/AISnakeCreator.java (hunk 4 above)
Comment (ester-agishtein) [reply src/main/java/touro/snake/AISnakeCreator.java:27]: Is there a way for me to put some type of delay though? Without any it looks like it's on fire.

File: src/main/java/touro/snake/Garden.java (hunk 8 above)
Comment (ester-agishtein) [reply src/main/java/touro/snake/Garden.java:41]: So when I change this, (and this is a question just in general) sometimes the aisnake will go off the board. But we're going to have an infinite board. Is that bug something I should fix or just leave?

File: src/main/java/touro/snake/AISnakeCreator.java [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+    public void moveSelf(Snake aiSnake) {
+
+        if (aiSnake.inBounds() && !aiSnake.eatsSelf()) {
+            Direction direction = this.randomDirection();
+            aiSnake.turnTo(direction);
+        }
+    }
+
+    private Direction randomDirection() {
+        int pick = new Random().nextInt(Direction.values().length);
```
Comment (andrewoid) [inline src/main/java/touro/snake/AISnakeCreator.java:21]: Do not create a new `Random`.
Create it once and reuse it.

File: src/main/java/touro/snake/Garden.java [hunk 10]
```diff
@@ -12,16 +12,23 @@
     private final Snake snake;
     private final FoodFactory foodFactory;
     private Food food;
+    private final Snake aiSnake;
 
-    public Garden(Snake snake, FoodFactory foodFactory) {
+    public Garden(Snake snake, Snake aiSnake, FoodFactory foodFactory) throws InterruptedException {
```
Comment (andrewoid) [inline src/main/java/touro/snake/Garden.java:17]: Why are you throwing `InterruptedException`

File: src/main/java/touro/snake/Garden.java [hunk 11]
```diff
@@ -12,16 +12,23 @@
...
         this.snake = snake;
         this.foodFactory = foodFactory;
+        this.aiSnake = aiSnake;
+        AISnakeCreator aiSnakeCreator = new AISnakeCreator(aiSnake);
```
Comment (andrewoid) [inline src/main/java/touro/snake/Garden.java:21]: What does this line do?

File: src/main/java/touro/snake/AISnakeCreator.java [hunk 12]
```diff
@@ -0,0 +1,25 @@
...
+
+public class AISnakeCreator {
+    private final Snake aiSnake;
+
+    public AISnakeCreator(Snake aiSnake) {
+        this.aiSnake = aiSnake;
+        moveSelf(aiSnake);
+    }
+
+    public void moveSelf(Snake aiSnake) {
```
Comment (andrewoid) [inline src/main/java/touro/snake/AISnakeCreator.java:12]: This doesn't actually move the Snake

File: src/main/java/touro/snake/AISnakeCreator.java [hunk 13]
```diff
@@ -0,0 +1,25 @@
+package touro.snake;
+import java.util.Random;
+
+public class AISnakeCreator {
```
Comment (andrewoid) [inline src/main/java/touro/snake/AISnakeCreator.java:4]: What you call your classes matters.
If I saw a class named `AISnakeCreator` I would think that the class's purpose was to create the AI Snake. That's not what this class is doing.

File: src/main/java/touro/snake/Garden.java (hunk 8 above)
Comment (andrewoid) [reply src/main/java/touro/snake/Garden.java:41]: What stops the AI Snake from moving?

**Pull request closed without merging**